collectors/node/metrics.go

```diff
@@ -15,225 +15,107 @@
+	NET_OUT         = "network.out"
+	NET_IN_PACKETS  = "network.in.packets"
+	NET_OUT_PACKETS = "network.out.packets"
+	NET_IN_DROPPED  = "network_in_dropped"
+	NET_OUT_DROPPED = "network_out_dropped"
+	NET_IN_ERRORS   = "network_in_errors"
+	NET_OUT_ERRORS  = "network_out_errors"
+
+	/* memory.<namespace> */
+	MEM_TOTAL   = "memory.total"
+	MEM_FREE    = "memory.free"
+	MEM_BUFFERS = "memory.buffers"
+	MEM_CACHED  = "memory.cached"
+
+	/* swap.<namespace> */
+	SWAP_TOTAL = "swap.total"
+	SWAP_FREE  = "swap.free"
+	SWAP_USED  = "swap.used"
+)
 
-type lastCPUTimes struct {
-	sync.Mutex
-	times cpu.TimesStat
+type nodeCollector struct {
+	datapoints []producers.Datapoint
 }
 
-var lastCPU lastCPUTimes
-
-func init() {
-	t, _ := cpu.Times(false) // get totals, not per-cpu stats
-	lastCPU.Lock()
-	lastCPU.times = t[0]
-	lastCPU.Unlock()
+type nodeMetricPoller interface {
+	poll() error
+	getDatapoints() ([]producers.Datapoint, error)
 }
 
-func getUptime() uint64 {
-	uptime, _ := host.Uptime()
-	return uptime
-}
+func getNodeMetrics() ([]producers.Datapoint, error) {
+	nc := nodeCollector{}
 
-func getProcessCount() uint64 {
-	i, _ := host.Info()
-	return i.Procs
-}
-
-func getNumCores() int32 {
-	cores := int32(0)
-	cpus, _ := cpu.Info()
-	for _, c := range cpus {
-		cores += c.Cores
+	nodeMetricPollers := []nodeMetricPoller{
+		&uptimeMetric{},
+		&cpuCoresMetric{},
+		&loadMetric{},
+		&filesystemMetrics{},
+		&memoryMetric{},
+		&networkMetrics{},
+		&processMetrics{},
 	}
-	return cores
-}
-
-func getLoadAvg() *load.AvgStat {
-	l, _ := load.Avg()
-	return l
-}
-
-func getCPUTimes() (cpu.TimesStat, cpu.TimesStat) {
-	currentTimes, _ := cpu.Times(false) // get totals, not per-cpu stats
-	lastTimes := lastCPU.times
-
-	lastCPU.Lock()
-	lastCPU.times = currentTimes[0] // update lastTimes to the currentTimes
-	lastCPU.Unlock()
-
-	return currentTimes[0], lastTimes
-}
-
-func getMemory() *mem.VirtualMemoryStat {
-	m, _ := mem.VirtualMemory()
-	return m
-}
-
-func getSwap() *mem.SwapMemoryStat {
-	s, _ := mem.SwapMemory()
-	return s
-}
 
-func getFilesystems() []nodeFilesystem {
-	f := []nodeFilesystem{}
-	parts, _ := disk.Partitions(false) // only physical partitions
-	for _, part := range parts {
-		usage, _ := disk.Usage(part.Mountpoint)
-		f = append(f, nodeFilesystem{
-			Name:               usage.Path,
-			CapacityTotalBytes: usage.Total,
-			CapacityUsedBytes:  usage.Used,
-			CapacityFreeBytes:  usage.Free,
-			InodesTotal:        usage.InodesTotal,
-			InodesUsed:         usage.InodesUsed,
-			InodesFree:         usage.InodesFree,
-		})
+	// For each metric poller defined, execute .poll() to get the latest
+	// metric, check for errors, and iterate over the slice of producers.Datapoint
+	// returned by .poll(), adding them to our top scope nodeMetrics slice.
+	for _, mp := range nodeMetricPollers {
+		if err := mp.poll(); err != nil {
+			return nc.datapoints, err
+		}
+
+		if dps, err := mp.getDatapoints(); err != nil {
```

Comment:
it looks like you can join `poll` and `getDatapoints` into one function, i don't see much value in having 2 separate functions. Considering the amount of work this might take to change the interface and tight timeframe, this can be addressed in the follow up PR